### DISCUSSION TITLE
EM_JS is not supported in side modules

### DISCUSSION THREAD

**Discussion by aothms** [pyodide/pyodide#3229] [Q&A]
Dear all, thank you for your work on pyodide. I'm really enthusiastic about the possibilities this provides.

I'm the maintainer of a python module and setting my first steps in wasm land. So please bear with me. My module is a bit of a beast to compile. We have a build scripts but no setup.py, so pywasmcross wasn't of much help (afaict).

Our build script compiles all dependencies as static modules, which is then compiled into one dynamic python extension library. Prefixing most of the calls with `emconfigure` and friends I have now arrived at some issues.

**Fixed:**

1. `need the dylink section to be first` fixed with `-s SIDE_MODULE` thanks: https://github.com/pyodide/pyodide/pull/2637#issuecomment-1140345077
2. that caused many `duplicate symbol` errors (because of whole-archive), I managed to fixed that by removing the repeated static libraries (poor man's start-group end-group) (`-Wl,--allow-multiple-definition` wasn't an option as it's not supported in `wasm-ld`)

**Issue:**

3. **now I've ran into the issue that `EM_JS is not supported in side modules`**

I first tried explicitly setting `-s WASM` but that appears to be the default already.

The error is triggered at:

https://github.com/emscripten-core/emscripten/blob/main/emscripten.py#L342

Inspecting the `emJsFuncs`:

https://github.com/emscripten-core/emscripten/blob/main/tools/extract_metadata.py#L338

```
{'OSD_MemInfo_getModuleHeapLength': '()<::>{ return Module.HEAP8.length; }',
 '_Py_CheckEmscriptenSignals_Helper': '(void)<::>{ if '
                                      '(!Module.Py_EmscriptenSignalBuffer) { '
                                      'return 0; } try { let result = '
                                      'Module.Py_EmscriptenSignalBuffer[0]; '
                                      'Module.Py_EmscriptenSignalBuffer[0] = '
                                      '0; return result; } catch(e) { return '
                                      '0; } }',
 'em_call_init_function': '(PyInitFunctionType f)<::>{ return '
                          'Module.wasmTable.get(f)(); }',
 'getter_call_trampoline': '(getter get, PyObject *obj, void *closure)<::>{ '
                           'return wasmTable.get(get)(obj, closure); }',
 'method_call_trampoline': '(PyCFunction func, PyObject *self, PyObject *arg1, '
                           'PyObject *arg2)<::>{ return '
                           'wasmTable.get(func)(self, arg1, arg2); }',
 'occJSConsoleDebug': '(const char* theStr)<::>{ '
                      'console.debug(UTF8ToString(theStr)); }',
 'occJSConsoleError': '(const char* theStr)<::>{ '
                      'console.error(UTF8ToString(theStr)); }',
 'occJSConsoleInfo': '(const char* theStr)<::>{ '
                     'console.info(UTF8ToString(theStr)); }',
 'occJSConsoleWarn': '(const char* theStr)<::>{ '
                     'console.warn(UTF8ToString(theStr)); }',
 'setter_call_trampoline': '(setter set, PyObject *obj, PyObject *value, void '
                           '*closure)<::>{ return wasmTable.get(set)(obj, '
                           'value, closure); }'}
```

* The OSD_ and occJS ones seem to come from OpenCASCADE which has specific code paths selected using the preprocessor. I can patch them out I think if needed https://github.com/Open-Cascade-SAS/OCCT/blob/master/src/OSD/OSD_MemInfo.cxx#L40

* The others seem to come from **pyodide itself** https://github.com/pyodide/pyodide/blob/stable/cpython/patches/0001-Patch-in-call-trampolines-to-handle-fpcast-troubles.patch#L82 https://github.com/pyodide/pyodide/blob/stable/cpython/patches/0008-Fix-_PyImport_LoadDynamicModuleWithSpec-fpcast.patch#L23

My suspicion is that I'm probably missing some compiler/linker flag. I hope somebody can advise.

**Comment by hoodmane** [comment]
The emjs calls from Pyodide are probably appearing due to incorrect linking. You shouldn't be statically linking libpython into the side modules.

The js console functions should use [emscripten_log](https://emscripten.org/docs/api_reference/emscripten.h.html#c.emscripten_log). `OSD_MemInfo_getModuleHeapLength` can be done using [`emscripten_run_script_int`](https://emscripten.org/docs/api_reference/emscripten.h.html#c.emscripten_run_script_int), but the original approach is better if in the main module since in that case you can run even in web pages that do not allow unsafe dynamic eval. This can be put behind a `#if SIDE_MODULE` conditional.

**Comment by aothms** [comment]
Many thanks for the quick and insightful reply.

1)

Patching the code quickly I found this note on emscriptem_log()

> don't use bogus emscripten_log() corrupting UNICODE strings

https://github.com/Open-Cascade-SAS/OCCT/blob/master/src/Message/Message_PrinterSystemLog.cxx#L172

Anyway I don't think it affects me, so I just backspaced the emscripten related bits from their code.

No reply necessary.

2)

> You shouldn't be statically linking libpython into the side modules.

Thanks, that makes sense. My code uses various python symbols, what's the proper way then to handle the undefined symbols at link time? `-Wl,undefined,suppress` works, but is there a more specific approach?

3)

With your help, my module appears to link correctly, but loading still fails. I know get a somewhat ominous:

Firefox:

> CompileError: wasm validation error: at offset 14891: too many imports

Chrome:

> CompileError: WebAssembly.instantiate(): imports count of 170408 exceeds internal limit of 100000

Is this a hard limit? I'm not too versed in compiler terminology. Would you happen to know what an `import` is in this case? Is this something to address with `-fvisibility=hidden` (I disabled that because `--gc-sections` didn't work either) or is there an equivalent of `strip` for wasm binaries? Or does this need to be solved in the code? Or is this more likely an error hinting at corruption somewhere? Sorry this is all rather new to me,

**Comment by aothms** [comment]
Answering some of my own questions:

While wasm-strip (from https://github.com/WebAssembly/wabt) did drastically cut down the file size. It caused the wasm to be no longer loaded, because it removed the `dylink` section. I didn't see any flags how it could be tamed a bit.

Following `pywasmcross`'s behaviour and specifying `-sSIDE_MODULE=2 -sEXPORTED_FUNCTIONS=PyInit_<module_name>` helped advance a bit. The wasm loads now, but I get an actual python ImportError about the PyInit function not being present.

It does seem to exist though in the dynamic library as:

    $ grep -aoE 'PyInit\w+' _ifcopenshell_wrapper.so
    PyInit__ifcopenshell_wrapper

**Comment by aothms** [comment]
Ok, the grep thing was a bit of a red herring. I think an underscore was needed in front of the symbol:

    -sEXPORTED_FUNCTIONS=_PyInit__...
                         ^

As noted here

> `// Note the necessary prefix of "_".`
> 
> https://github.com/emscripten-core/emscripten/blob/1.29.12/src/settings.js#L388](https://github.com/emscripten-core/emscripten/blob/1.29.12/src/settings.js#L392)

Unfortunately now I obtain the error

> LinkError: WebAssembly.instantiate(): Import #711 module="env" function="time" error: imported function does not match the expected type

Using wasm2c (https://github.com/WebAssembly/wabt/blob/main/wasm2c/README.md). I see this definition:

```
/* import: 'env' 'time' */
u64 Z_envZ_time(struct Z_env_instance_t*, u32);
```

Observing https://github.com/pyodide/pyodide/discussions/3100 I already compile with the `-sWASM_BIGINT` flag. Does the issue have to be related to that? I.e am I linking some in some bytecode by accident compiled without this flag, or could there be something else, like a symbol collision?

Hoping for some pointers. Thanks in advance.

**Comment by aothms** [comment]
Actually. This trivial module (just a wrapper around time()) exhibits the same problem. So either there is something trivially wrong with the way I have setup things. Or we're looking at sth more fundamental.

```
::::::::::::::
simple/CMakeLists.txt
::::::::::::::
project(mytime)
cmake_minimum_required(VERSION 3.22)
find_package(SWIG)
include(${SWIG_USE_FILE})

find_package(PythonLibs)
include_directories(${PYTHON_INCLUDE_DIRS})
include_directories(${CMAKE_CURRENT_SOURCE_DIR})

add_library(util SHARED mytime.cpp)

set_source_files_properties(mytime.i PROPERTIES CPLUSPLUS ON)
swig_add_library(mytime LANGUAGE python SOURCES mytime.i)
swig_link_libraries(mytime util)


::::::::::::::
simple/mytime.cpp
::::::::::::::
#include "mytime.h"

#include <ctime>

int get_time() {
    return (int) time(nullptr);
}


::::::::::::::
simple/mytime.h
::::::::::::::
int get_time();


::::::::::::::
simple/mytime.i
::::::::::::::
%module mytime

%{
#define SWIG_FILE_WITH_INIT
#include "mytime.h"
%}

%include "mytime.h"


::::::::::::::
simple/setup.py
::::::::::::::
from setuptools import setup

setup(name='mytime',
      version='0.1.0',
      description='',
      author='Thomas Krijnen',
      author_email='thomas@aecgeeks.com',
      packages=['mytime'],
      package_dir={'mytime': 'build'},
      package_data={'': ['*.so']},
)


::::::::::::::
simple.yaml
::::::::::::::
package:
  name: mytime
  version: 0.1.0

source:
  path: simple

build:
  script: |
    mkdir build && cd build
    CXXFLAGS="-sSIDE_MODULE -sWASM_BIGINT" \
    CFLAGS="-sSIDE_MODULE -sWASM_BIGINT" \
    LDFLAGS="-sSIDE_MODULE -sWASM_BIGINT" \
    emcmake cmake .. -DPYTHON_INCLUDE_DIR=/src/pyodide/cpython/installs/python-3.10.2/include/python3.10
    make
    touch __init__.py

about:
  home: n/a
  license: n/a
  summary: n/a
```

Result: 

> pyodide.JsException: LinkError: WebAssembly.instantiate(): Import #0 module="env" function="time" error: imported function does not match the expected type

**Reply by hoodmane** [reply]
If you upload these files as a zip archive I can take a look. `wasm-objdump` from the wasm binary toolkit will probably be helpful here.

**Reply by aothms** [reply]
That would be highly appreciated. I put the files in a repo along with start to end instructions of how I built the whl. https://github.com/aothms/pyodide-pkg-with-time Looking forward to your comments :)

**Reply by hoodmane** [reply]
Well I looked into this and I am totally confused. 

In a loaded version of Pyodide I see:
```js
>>> WebAssembly.Function.type(pyodide._module.asm.time)
{"parameters":["i32"],"results":["i32"]}
```
inspecting `pyodide.asm.wasm` to see where `time` comes from we see:
```sh
$ wasm-objdump pyodide.asm.wasm -j Export -x | grep '\btime\b'
 - func[7618] <__time> -> "time" 
```
so the `time` function is the `__time` function defined in `system/lib/libc/emscripten_time.c`. It is declared with signature `time_t __time(time_t *t)`. We have a 64 bit `time_t` and 32 bit pointers, so the signature should be `(i32) -> i64` not `(i32) -> i32` which is reported by `WebAssembly.Function.type`.

Using `wasm-objdump` to investigate the Pyodide webassembly module we see that it declares the function to be of type `(i32) -> i64` like we expect:
```sh
$ wasm-objdump pyodide.asm.wasm -j Function -x | rg '\b__time\b'
 - func[7618] sig=20 <__time>
```
This says that `__time` has type 20. We also have to look up what that means.
```
$ wasm-objdump pyodide.asm.wasm -j Type -x | grep '\[20\]'
 - type[20] (i32) -> i64
```
it looks like `time` is defined to be `(i32) -> i64` in the webassembly module! Exactly what we expect. Why does it look different at runtime?

Inspecting your `libutil.so` module:
```sh
$ wasm-objdump libutil.so -j Import -x | grep 'time'
 - func[0] sig=0 <env.time> <- env.time
$ wasm-objdump libutil.so -j Type -x | grep '\[0\]'
 - type[0] (i32) -> i64
```
So `libutil.so` thinks the type of `time` is what I think it should be.

**Reply by hoodmane** [reply]
Never mind the problem is a simple version mismatch:
```js
<script type="text/javascript" src="https://cdn.jsdelivr.net/pyodide/v0.21.3/full/pyodide.js"></script>
```
In between v0.21.3 and the tip of tree, `time_t` changed from 32 bit to 64 bit. If you change the url to `v0.22.0a1` the webassembly module loads fine and you get:
```js
ImportError: dynamic module does not define module export function (PyInit__mytime)
```

**Reply by hoodmane** [reply]
This next problem is due to the fact that `_mytime.so` never finishes loading for some reason:
```js
> pyodide._module.LDSO.loadedLibsByName['/lib/python3.10/site-packages/mytime/_mytime.so'].module
< 'loading'
```
If we switch to loading it with `pyodide.loadPackage` instead of `micropip`:
```js
await pyodide.loadPackage("dist/mytime-0.1.0-py3-none-any.whl");
```
we get:
```js
Failed to load mytime
pyodide.asm.js:5632 The following error occurred while loading mytime:
loadPackage @ pyodide.asm.js:5632
await in loadPackage (async)
main @ (index):19
await in main (async)
(anonymous) @ (index):25
pyodide.asm.js:5635 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'message')
    at Object.loadPackage (pyodide.asm.js:5635:35)
    at async main ((index):19:9)
```
So for one thing we discovered a bug in micropip: it swallowed that error. @ryanking13

**Reply by hoodmane** [reply]
Uggggh something is throwing undefined. Why is `throw undefined` even allowed in JavaScript??

**Reply by hoodmane** [reply]
Okay, now the problem is that it's looking for `_mytime.so` in the same directory rather than in one of the standard library directories. This can be fixed with the following patch:
```patch
--- a/src/js/dynload.ts
+++ b/src/js/dynload.ts
@@ -29,10 +29,11 @@ export async function loadDynlib(lib: string, shared: boolean) {
     console.debug(`Loading a dynamic library ${lib}`);
   }
 
+  const libDir = lib.substring(0, lib.lastIndexOf("/"));
   // This is a fake FS-like object to make emscripten
   // load shared libraries from the file system.
   const libraryFS = {
-    _ldLibraryPaths: ["/usr/lib", API.sitepackages],
+    _ldLibraryPaths: ["/usr/lib", API.sitepackages, libDir],
     _resolvePath: (path: string) => {
       if (DEBUG) {
         console.debug(`Searching a library from ${path}, required by ${lib}.`);
```
@ryanking13 this might be a good idea?

The next trouble is `_mytime.so` expected `libutil.so` to be loaded globally. The existing solutions to are either to use `dlopen` and `dlsym` or to make `libutil.so` into a separate `sharedlibrary` package. We automatically load shared libraries as global. If I add an extra `__init__.py` file to force it to load `libutil.so` as a global DSO:
```py
from .mytime import get_time
from pyodide_js._module import loadDynamicLibrary
loadDynamicLibrary("libutil.so", **{'global': True})
del mytime
```
 then it all works:
```js
> pyodide.runPython("mytime.get_time()")
< 1667763585
```

**Reply by hoodmane** [reply]
@ryanking13 I wonder: maybe we should make packages themselves declare a list of dynlibs they need loaded globally. Then we could load all sharedlibs as not global by default and after doing the loading for the libs do a second pass where we turn the ones that need to be global into global libs. Because this is an example where `libutil.so` is in the same package but it needs to be global. Whether a library needs to be loaded globally is really more a property of a consumer than a property of the provider.

We maybe should also make a public function to make an already loaded library global to give people a work around if they are using wheels that don't have the manifest info in `repodata.json`... Or maybe `loadPackage` and/or `micropip.install` could take an extra argument to indicate this.

**Reply by aothms** [reply]
Wow, this is some next level sorcery. Thank you. I can confirm that using v0.22.0a1 and switching libutil from shared to static, solves this issue without needing to patch anything. Distributing shared libraries as individually installable packages also makes a lot of sense in the long run. Really excited now to try my real module...

**Reply by ryanking13** [reply]
> @ryanking13 I wonder: maybe we should make packages themselves declare a list of dynlibs they need loaded globally. Then we could load all sharedlibs as not global by default and after doing the loading for the libs do a second pass where we turn the ones that need to be global into global libs. Because this is an example where libutil.so is in the same package but it needs to be global.

Yeah, that is what I am going to add in #3183, but not by declaring a list of global libs but in a smart way by checking metadata inside the wasm module.
